fix: owner changing ownership causing identical previous and current ids when using a distributed authority network topology

NoelStephensUnity opened the pull request:
This fixes an issue with the synchronization of NetworkVariables when changing ownership where it was possible to have the previous owner be equal to the current owner when using a distributed authority network topology and the owner handled changing the ownership.

This includes an additional check for dirty states of any collections base NetworkVariables.

<!-- Add short version of the JIRA ticket to the PR title (e.g. "feat: new shiny feature [MTT-123]") -->

fix: #3343 
close: #3343 

## Changelog

- Fixed issue where an owner that changes ownership, when using a distributed authority network topology, could yield identical previous and current owner identifiers. This could also cause `NetworkTransform` to fail to change ownership which would leave the previous owner still subscribed to network tick events.

## Testing and Documentation

- Includes integration test `NetworkObjectOwnershipTests.TestAuthorityChangingOwnership`.
- No documentation changes or additions were necessary.

<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

NoelStephensUnity committed:
fix

This fixes an issue with the synchronization of NetworkVariables when changing ownership where it was possible to have the previous owner be equal to the current owner when using a distributed authority network topology and the owner handled changing the ownership.

This includes an additional check for ditry states of any collections base NetworkVariables.

This bug was introduced as a regression bug when resolving the issue that any owner read permission NetworkVariables were fully synchronized with the new owner.

NoelStephensUnity committed:
update

adding a change log entry.

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=3347) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=3347) before we can accept your contribution.<br/>**5** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: NoelStephensUnity<br/>:white_check_mark: EmandM<br/>:white_check_mark: DSchroer<br/>:white_check_mark: fluong6<br/>:white_check_mark: michalChrobot<br/>:x: simon-lemay-unity<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=3347) it.</sub>

NoelStephensUnity committed:
fix

Fixing issue where we need to reset the OwnerId and PreviousOwnerId after running through NetworkBehaviourUpdate to assure the values are properly set when there are no dirty objects to reset.

NoelStephensUnity committed:
update

Updating change log entry and adding PR number.

NoelStephensUnity committed:
fix

Fixing an order of operations issue with synchronizing owner read permission NetworkVariables.

NoelStephensUnity committed:
test

adding a test to validate this specific issue where the authority changes the ownership.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/da-change-ownership-results-in-invalid-previous-and-current-clientids

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/da-change-ownership-results-in-invalid-previous-and-current-clientids

EmandM commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -432,6 +432,15 @@ internal void RemoveOwnership(NetworkObject networkObject)
+            {
+                if (NetworkManager.LogLevel <= LogLevel.Developer)
+                {
+                    Debug.LogWarning($"[{nameof(NetworkSpawnManager)}][{nameof(ChangeOwnership)}] Attempting to change ownership to Client-{clientId} when the owner is already {networkObject.OwnerClientId}! (Ignoring)");
```
This check is already happening on [line 524](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/eca5b2744a162998b006ea98394334762d54e6f7/com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs#L524) of this function, though the check is a little different:

``` C#
if (networkObject.OwnerClientId == clientId && networkObject.PreviousOwnerId == clientId)
```
We should probably consolidate the two checks.

NoelStephensUnity replied:
The issue was that there are several places that invoke `NetworkSpawnManager.ChangeOwnership`:
![image](https://github.com/user-attachments/assets/75aac9e8-cab9-4126-be35-4aec36af54a4)
I decided to put a check at the funnel point. However, at some point in the future we might look at all of the invocations to determine if it makes sense to consolidate the various checks for this condition... for now we know that we will at least see warnings...maybe I should make that an error so it will fail any PRs that introduce this condition?

EmandM replied:
I think the bigger thing is that the check is happening twice in the same function. So the same code path is going to check the same check twice.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -530,6 +539,10 @@ internal void ChangeOwnership(NetworkObject networkObject, ulong clientId, bool
                 return;
             }
 
+            // Save the previous owner to work around a bit of a nasty issue with assuring NetworkVariables are serialized when ownership changes
```
Do we have any details on this issue?

NoelStephensUnity replied:
We have that document for near future tasks... where changing ownership does a synchronization (for now it does a full synch of owner read and write NetworkVariables.

This section of code actually was one of the places that was causing the issue for this PR.
[Look down below at line 564](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/eca5b2744a162998b006ea98394334762d54e6f7/com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs#L564)
Those two variables are used.

The issue is that we have to save off the original previous owner id and the original owner id in order to be able to run through the standard `NetworkVariable` synchronization process where `NetworkVariableDeltaMessage`s will be generated and sent prior to the change in ownership message.

What sits between storing those properties into locals and the call to `networkObject.SynchronizeOwnerNetworkVariables` is this:
```C#
            // Used to distinguish whether a new owner should receive any currently dirty NetworkVariable updates
            networkObject.PreviousOwnerId = networkObject.OwnerClientId;

            // Assign the new owner
            networkObject.OwnerClientId = clientId;

            // Always notify locally on the server when ownership is lost
            networkObject.InvokeBehaviourOnLostOwnership();

            // Authority adds entries for all client ownership
            UpdateOwnershipTable(networkObject, networkObject.OwnerClientId);

            // Always notify locally on the server when a new owner is assigned
            networkObject.InvokeBehaviourOnGainedOwnership();
```
Which we have to maintain the order of operations for users (i.e. the previous and new owner should be set prior to notifying ownership is lost, the ownership table is updated, and ownership is gained... once that is done we sort of "rollback" to the previous original owner so when it is doing checks to make sure the owner can read and write (prior to determining if it can serialize and send deltas) any NetworkVariables associated with the NetworkBehaviour....
*eyes crossing*

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/da-change-ownership-results-in-invalid-previous-and-current-clientids

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/da-change-ownership-results-in-invalid-previous-and-current-clientids

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/da-change-ownership-results-in-invalid-previous-and-current-clientids
